File: ml-benchmark/synduce/tree/minmax.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};

max = \a: Int. \b: Int. if < a b then b else a;
min = \a: Int. \b: Int. if > a b then b else a;

g = fix (
  \f: {Int, Int} -> Tree -> {Int, Int}. \x: {Int, Int}. \t: Tree.
  match t with
    leaf a -> {min a x.1, max a x.2}
  | node {a, l, r} -> f (f {min a x.1, max a x.2} l) r
  end
);

spec = fix (
  \f: Tree -> {Int, Int}. \t: Tree.
  match t with
    leaf x -> {x,x}
  | node {a, l, r} ->
    let result = g {a, a} l in
    g {result.1, result.2} r
  end
);

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    leaf a -> leaf a
  | node {a, l, r} -> node {a, f l, f r}
  end
);

main = \t: Tree. spec (repr t);

File: ml-benchmark/synduce/tree/minmax.ml
type tree = Leaf of int | Node of int * tree * tree

let max a b = if a < b then b else a
let min a b = if a > b then b else a

let rec g x t =
  match t with
  | Leaf a ->
      (match x with
       | (x1, x2) -> (min a x1, max a x2))
  | Node (a, l, r) ->
      (match x with
       | (x1, x2) -> g (g (min a x1, max a x2) l) r)

let rec spec t =
  match t with
  | Leaf x -> (x, x)
  | Node (a, l, r) ->
      match g (a, a) l with
      | (r1, r2) -> g (r1, r2) r

let rec repr t =
  match t with
  | Leaf a -> Leaf a
  | Node (a, l, r) -> Node (a, repr l, repr r)

let program t = spec (repr t)